src/Databases/PostgreSQL/FetchFromPostgreSQL.cpp

```diff
@@ -0,0 +1,118 @@
+#include <Databases/PostgreSQL/FetchFromPostgreSQL.h>
+
+#if USE_LIBPQXX
+
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/trim.hpp>
+#include <pqxx/pqxx>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNKNOWN_TYPE;
+}
+
+
+std::shared_ptr<NamesAndTypesList> fetchTableStructure(ConnectionPtr connection, const String & postgres_table_name, bool use_nulls)
+{
+    auto columns = NamesAndTypesList();
+
+    std::string query = fmt::format(
+           "SELECT attname AS name, format_type(atttypid, atttypmod) AS type, "
+           "attnotnull AS not_null, attndims AS dims "
+           "FROM pg_attribute "
+           "WHERE attrelid = '{}'::regclass "
+           "AND NOT attisdropped AND attnum > 0", postgres_table_name);
+    pqxx::read_transaction tx(*connection);
+    pqxx::stream_from stream(tx, pqxx::from_query, std::string_view(query));
+    std::tuple<std::string, std::string, std::string, uint16_t> row;
+
+    while (stream >> row)
+    {
+        columns.push_back(NameAndTypePair(
+                std::get<0>(row),
+                getDataType(std::get<1>(row), use_nulls && (std::get<2>(row) == "f"), std::get<3>(row))));
+    }
+    stream.complete();
+    tx.commit();
+
+    if (columns.empty())
+        return nullptr;
+
+    return std::make_shared<NamesAndTypesList>(columns);
+}
+
+
+DataTypePtr getDataType(std::string & type, bool is_nullable, uint16_t dimensions)
+{
+    DataTypePtr res;
+
+    /// Get rid of trailing '[]' for arrays
+    if (dimensions)
+        type.resize(type.size() - 2);
+
+    if (type == "smallint")
+        res = std::make_shared<DataTypeInt16>();
+    else if (type == "integer")
+        res = std::make_shared<DataTypeInt32>();
+    else if (type == "bigint")
+        res = std::make_shared<DataTypeInt64>();
+    else if (type == "real")
+        res = std::make_shared<DataTypeFloat32>();
+    else if (type == "double precision")
+        res = std::make_shared<DataTypeFloat64>();
+    else if (type == "serial")
+        res = std::make_shared<DataTypeUInt32>();
+    else if (type == "bigserial")
+        res = std::make_shared<DataTypeUInt64>();
+    else if (type.starts_with("timestamp"))
+        res = std::make_shared<DataTypeDateTime>();
+    else if (type == "date")
+        res = std::make_shared<DataTypeDate>();
+    else if (type.starts_with("numeric"))
```

Comment:
BTW, our DataTypeFactory supports most of the cases automatically.
And it can parse `numeric(x, y)` as well.

Reply:
> And it can parse numeric(x, y) as well.

Done.